[AIRFLOW-255] schedule_dag shouldn't return early if dagrun_timeout is given.

reconditesea opened the pull request:
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- https://issues.apache.org/jira/browse/AIRFLOW-255
- Testing Done:  ./run_unit_tests.sh

In https://github.com/apache/incubator-airflow/blob/master/airflow/jobs.py#L396, schedule_dag returns as long as `len(active_runs) >= dag.max_active_runs`. It should also take into account dag.dagrun_timeout

bolkedebruin commented on airflow/jobs.py:
```diff
@@ -393,14 +393,19 @@ def schedule_dag(self, dag):
                 session=session
             )
-            if len(active_runs) >= dag.max_active_runs:
+            # retunr if already reach maximum active runs and no timeout setting
```
nit: typo

bolkedebruin commented on airflow/jobs.py:
```diff
@@ -393,14 +393,19 @@ def schedule_dag(self, dag):
                         dr.start_date < datetime.now() - dag.dagrun_timeout):
                     dr.state = State.FAILED
                     dr.end_date = datetime.now()
+                    timeout_triggered += 1
```
please rename this variable. "failed_runs" seems better to me

reconditesea replied:
How about timedout_runs? As timed out doesn't necessarily mean failure.

bolkedebruin replied:
Also fine!

bolkedebruin commented on airflow/jobs.py:
```diff
@@ -393,14 +393,19 @@ def schedule_dag(self, dag):
+                    timeout_triggered += 1
+            if len(active_runs) - timeout_triggered >= dag.max_active_runs:
+                return
             session.commit()
```
the commit needs to happen before returning, now your losing the state update.

bolkedebruin commented on tests/jobs.py:
```diff
@@ -685,6 +685,45 @@ def test_scheduler_fail_dagrun_timeout(self):
+        dr = scheduler.schedule_dag(dag)
+        self.assertIsNotNone(dr)
+
+        # Should not be scheduled as no dagrun_timeout specified
```
Nit: I think this assertion is incorrect: ie. should be "Should not be scheduled as DagRun has not timedout and max_active_runs is reached". It is now somewhat confusing.

reconditesea replied:
Nice catch!

bolkedebruin commented on tests/jobs.py:
```diff
@@ -685,6 +685,45 @@ def test_scheduler_fail_dagrun_timeout(self):
+        new_dr = scheduler.schedule_dag(dag)
+        self.assertIsNone(new_dr)
+
+        # Should be scheduled as dagrun_timeout is passed
```
nit: is = has

bolkedebruin commented on airflow/jobs.py:
```diff
@@ -393,15 +393,20 @@ def schedule_dag(self, dag):
                 session=session
             )
-            if len(active_runs) >= dag.max_active_runs:
+            # return if already reach maximum active runs and no timeout setting
```
nit: reach = reached

reconditesea committed:
[ARFLOW-255] schedule_dag method shouldn't return early if dagrun_timeout is given.

    In https://github.com/apache/incubator-airflow/blob/master/airflow/jobs.py#L396, schedule_dag returns as long as len(active_runs) >= dag.max_active_runs.
    It should also take into account dag.dagrun_timeout
